pkg/metricsadapter/provider/custommetrics.go

```diff
@@ -3,33 +3,285 @@ package provider
 import (
 	"context"
 	"fmt"
+	"reflect"
+	"strings"
 
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	listcorev1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
 	"k8s.io/metrics/pkg/apis/custom_metrics"
+	custommetricsv1beta2 "k8s.io/metrics/pkg/apis/custom_metrics/v1beta2"
+	custommetricsclient "k8s.io/metrics/pkg/client/custom_metrics"
+	custommetricsschema "k8s.io/metrics/pkg/client/custom_metrics/scheme"
 	"sigs.k8s.io/custom-metrics-apiserver/pkg/provider"
+
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	clusterlister "github.com/karmada-io/karmada/pkg/generated/listers/cluster/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+var (
+	versionConverter = custommetricsclient.NewMetricConverter()
 )
 
 // CustomMetricsProvider is a custom metrics provider
 type CustomMetricsProvider struct {
+	// newDiscoveryClientFunc returns a discovery client for member cluster apiserver
+	newDiscoveryClientFunc func(string) (*discovery.DiscoveryClient, error)
+	clusterLister          clusterlister.ClusterLister
 }
 
 // MakeCustomMetricsProvider creates a new custom metrics provider
-func MakeCustomMetricsProvider() *CustomMetricsProvider {
-	return &CustomMetricsProvider{}
+func MakeCustomMetricsProvider(clusterLister clusterlister.ClusterLister, secretLister listcorev1.SecretLister) *CustomMetricsProvider {
+	return &CustomMetricsProvider{
+		clusterLister:          clusterLister,
+		newDiscoveryClientFunc: discoveryClientForClusterFunc(clusterLister, secretLister),
+	}
 }
 
 // GetMetricByName will query metrics by name from member clusters and return the result
 func (c *CustomMetricsProvider) GetMetricByName(ctx context.Context, name types.NamespacedName, info provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValue, error) {
-	return nil, fmt.Errorf("karmada-metrics-adapter still not implement it")
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+	metricValueList := &custom_metrics.MetricValueList{}
+	for _, cluster := range clusters {
+		metrics, err := c.getMetricByName(ctx, cluster.Name, name, info, metricSelector)
+		if err != nil {
+			klog.Warningf("query %s's %s metric from cluster %s failed, err: %+v", info.GroupResource.String(), info.Metric, cluster.Name, err)
+			continue
+		}
+		metricValueList.Items = append(metricValueList.Items, metrics.Items...)
```

Comment:
Object is unique in one cluster but it may exist in multiple clusters. For this situation, we might add the value of all clusters.

Reply:
I am not sure if there are some metric values that are average type. @Poor12

Reply:
From the codebase of native hpa-controller, I think add is better. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/replica_calculator.go#L289

If users set `averageValue`, hpa-controller will use the following logic `averageUsage` * `podNum`.
```
usageRatio := float64(usage) / (float64(targetAverageUsage) * float64(replicaCount))
```
So the usage from the metrics-adapter might be the sum of all clusters.

Reply:
If you add metric values together, how are `Timestamp`, `WindowSeconds` and `DescribedObject` handled?

@Poor12

Reply:
The information is not actually used(only used for printing), so I think we can maybe just pick one.

Reply:
ok.